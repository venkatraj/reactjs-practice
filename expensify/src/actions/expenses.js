import uuid from 'uuid'

// Actions
const addExpense = (
  {
    description = '', 
    note = '', 
    amount = 0, 
    createdAt = 0
  } = {}
) => ({
  type: 'ADD_EXPENSE',
  expense: {
    id: uuid(),
    description,
    note,
    amount,
    createdAt  
  }
})

const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
})

const removeExpense = (id) => ({
  type: 'REMOVE_EXPENSE',
  id
})

export { addExpense, editExpense, removeExpense }