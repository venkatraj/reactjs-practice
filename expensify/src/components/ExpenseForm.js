import React from 'react';
import 'react-dates/initialize'
import { SingleDatePicker } from 'react-dates'
import 'react-dates/lib/css/_datepicker.css'
import moment from 'moment'



class ExpenseForm extends React.Component {

  constructor(props) {
    super(props)
    if (props.expense) {
      const { description, amount, note, createdAt } = props.expense
      this.state = {
        description,
        note,
        amount: (amount / 100).toString(),
        createdAt: moment(createdAt),
        calendarFocused: false,
        error: null,
        buttonText: 'Update Expense'
      }
    } else {
      this.state = {
        description: '',
        note: '',
        amount: '',
        createdAt: moment(),
        calendarFocused: false,
        error: null,
        buttonText: 'Add Expense'
      }
    }
  }


  onDescriptionChange = (e) => {
    const description = e.target.value
    this.setState(() => ({ description }))
  }

  onNoteChange = (e) => {
    const note = e.target.value
    this.setState(() => ({ note }))
  }

  onAmountChange = (e) => {
    const amount = e.target.value
    if (!amount || amount.match(/^\d+(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }))
    }
  }

  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }))
    }
  }

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }))
  }

  onSubmit = (e) => {
    e.preventDefault()

    if (!this.state.description || !this.state.amount) {
      const error = 'Please provided description and amount!'
      this.setState(() => ({ error }))
    } else {
      this.setState(() => ({ error: null }))
      const expense = {
        description: this.state.description,
        note: this.state.note,
        amount: parseFloat(this.state.amount) * 100,
        createdAt: this.state.createdAt.valueOf()
      }
      this.props.onSubmit(expense)
    }
  }

  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Description"
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            placeholder="Add a note to your expense (optional)"
            onChange={this.onNoteChange}
            value={this.state.note}
          >
          </textarea>
          <button>
            {this.state.buttonText}
          </button>
        </form>
      </div>
    )
  }
}

export default ExpenseForm