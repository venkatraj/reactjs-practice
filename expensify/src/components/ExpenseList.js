import React from 'react'
import { connect } from 'react-redux'
import ExpenseListItem from './ExpenseListItem'
import getVisibleExpenses from '../selectors/expenses'

export const ExpenseList = (props) => (
  <div>
    {
      props.expenses.length > 0 ? (
        props.expenses.map(expense => <ExpenseListItem key={expense.id} {...expense} />)
    ): (
      <p>No expenses</p>
      )
  }
  </div>
)

const mapStateToProps = (state) => ({
  expenses: getVisibleExpenses(state.expenses, state.filters)
})

export default connect(mapStateToProps)(ExpenseList)