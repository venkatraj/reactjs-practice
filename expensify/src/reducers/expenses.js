const expensesReducerDefaultState = []
const expensesReducer = (state = expensesReducerDefaultState, action) => {
  switch(action.type) {
    case 'ADD_EXPENSE':
      return state.concat({
        ...action.expense
      })
    case 'EDIT_EXPENSE':
      return state.map(expense => {
        if (action.id === expense.id) {
          return {
            ...expense,
            ...action.updates
          }
        } else {
          return expense
        }
      })
    case 'REMOVE_EXPENSE':
      return state.filter(expense => expense.id !== action.id)
    default:
      return state
  }
}

export default expensesReducer