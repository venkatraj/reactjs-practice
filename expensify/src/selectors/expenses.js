import moment from 'moment'

const getVisibleExpenses = (expenses, {text, sortBy, startDate, endDate}) => {
  return expenses.filter(expense => {
    const createdAtMoment = moment(expense.createdAt)
    const startDateMatch = startDate ? createdAtMoment.isSameOrAfter(startDate) : true
    const endDateMatch = endDate ? createdAtMoment.isSameOrBefore(endDate) : true
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase())
    return startDateMatch && endDateMatch && textMatch
  }).sort((a, b) => {
    if (sortBy === 'amount') {
      return a.amount > b.amount ? -1 : 1
    } else if (sortBy === 'date') {
      return a.createdAt > b.createdAt ? 1 : -1
    }
  })
}

export default getVisibleExpenses