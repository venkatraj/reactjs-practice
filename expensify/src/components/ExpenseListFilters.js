import React from 'react'
import { connect } from 'react-redux'
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters'
import 'react-dates/initialize'
import 'react-dates/lib/css/_datepicker.css'
import { DateRangePicker } from 'react-dates'



class ExpenseListFilters extends React.Component {
  state = {
    focusedInput: null
  }

  onDatesChange = ({ startDate, endDate }) => {
    this.props.dispatch(setStartDate(startDate))
    this.props.dispatch(setEndDate(endDate))
  }

  onFocusChange = (focusedInput) => {
    this.setState(() => ({ focusedInput }))
  }

  render() {
    return (
      <div>
        <input
          type='text'
          value={this.props.filters.text}
          onChange={e => this.props.dispatch(setTextFilter(e.currentTarget.value))}
        />
        <select
          value={this.props.filters.sortBy}
          onChange={e => {
            if (e.currentTarget.value == 'date') {
              this.props.dispatch(sortByDate())
            } else if (e.currentTarget.value == 'amount') {
              this.props.dispatch(sortByAmount())
            }
          }}
        >
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          startDateId="startDate"
          endDateId="endDate"
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.focusedInput}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
          showClearDates={true}
        />
      </div>
    )
  }
}

const mapStateToProps = (state) => {
  return {
    filters: state.filters
  }
}

export default connect(mapStateToProps)(ExpenseListFilters)