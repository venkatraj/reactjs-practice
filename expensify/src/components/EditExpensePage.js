import React from 'react'
import { connect } from 'react-redux'
import { editExpense, removeExpense } from '../actions/expenses'
import ExpenseForm from './ExpenseForm'

const EditExpensePage = (props) => (
  <div>
    <ExpenseForm
      expense={props.expense}
      onSubmit={expense => {
        props.dispatch(editExpense(props.expense.id, expense))
        props.history.push('/')
      }}
    />
    <button
      onClick={e => {
        props.dispatch(removeExpense(props.expense.id))
        props.history.push('/')
      }}
    >
      Remove
    </button>
  </div>
)

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id)
})


export default connect(mapStateToProps)(EditExpensePage)