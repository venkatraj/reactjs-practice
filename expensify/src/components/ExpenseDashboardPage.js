import React from 'react'
import ExpenseList from './ExpenseList'

const ExpenseDashboardPage = () => (
  <div>
    ExpenseDashboardPage
    <ExpenseList />
  </div>
)

export default ExpenseDashboardPage