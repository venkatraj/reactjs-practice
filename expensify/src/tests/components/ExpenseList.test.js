import React from 'react'
import { shallow } from 'enzyme'
import { ExpenseList } from '../../components/ExpenseList'
import expenses from '../fixtures/expenses'

test('Should render ExenseList with expenses', () => {
  const wrapper = shallow(<ExpenseList expenses={expenses} />)
  expect(wrapper).toMatchSnapshot()
})

test('Should render ExenseList with out expenses and empty message', () => {
  const wrapper = shallow(<ExpenseList expenses={[]} />)
  expect(wrapper).toMatchSnapshot()
})