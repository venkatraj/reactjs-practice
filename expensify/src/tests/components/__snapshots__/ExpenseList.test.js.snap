// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExenseList with expenses 1`] = `
<div>
  <ExpenseListItem
    amount={76500}
    createdAt={-345600000}
    description="Gas"
    id="1"
    key="1"
    note="For Cooking"
  />
  <ExpenseListItem
    amount={400000}
    createdAt={0}
    description="Rent"
    id="2"
    key="2"
    note="Final Payment"
  />
  <ExpenseListItem
    amount={2500}
    createdAt={345600000}
    description="Water"
    id="3"
    key="3"
    note=""
  />
</div>
`;

exports[`Should render ExenseList with out expenses and empty message 1`] = `
<div>
  <p>
    No expenses
  </p>
</div>
`;
